Process files by slices

nicolas-coulon opened the pull request:
The intent of this pull request is to process the input files by slices instead of reading them entirely at each poll time and whatever the monitor mode (i.e. : Update, Tail). This allows to deal with very large files.  Also, it solves issue #462.

An optional configuration parameter was provided for that : "connect.ftp.monitor.slicesize"

Previous version of code was kept and isolated, see comments "ENA: START previous version of code". 

Previous version of code is run if "connect.ftp.monitor.slicesize" is left blank.

nicolas-coulon committed:
add UpdateSlice mode

#328

nicolas-coulon committed:
add test + correct a bug

#328

nicolas-coulon committed:
add MaxLinesConverter + test

#328

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -53,77 +55,155 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+        case _ =>
+          logger.info(s" no fetch required")
+          false
+      }
```
I think the else & the match case could be combined. If I'm not wrong, most of the cases are identical ? 

Something like :

```scala
(mode, metadata) match {
    case (MonitorMode.UpdateSlice, Some(known) => ...
}
```

nicolas-coulon replied:
The idea is to remove (maybe in next commit), the legacy code. So I don't want to mix my code with the legacy code.

So this :
```
if( !isSliceMode ) {
      //ENA : legacy code
      metadata match {
        case None => logger.debug(s"${file.name} hasn't been seen before"); true
        case Some(known) if known.attribs.size != file.ftpFile.getSize => {
          logger.debug(s"${file.name} size changed ${known.attribs.size} => ${file.size}")
          true
        }
        case Some(known) if known.attribs.timestamp != file.timestamp => {
          logger.debug(s"${file.name} is newer ${known.attribs.timestamp} => ${file.timestamp}")
          true
        }
        case _ => false
      }
    } else {
      metadata match {
        case None =>
          logger.info(s"${file.name} hasn't been seen before")
          true
        case Some(previous) if previous.attribs.size != file.ftpFile.getSize => {
          logger.info(s"${file.name} size changed ${previous.attribs.size} => ${file.size}")
          true
        }
        case Some(previous) if previous.attribs.timestamp != file.timestamp => {
          logger.info(s"${file.name} is newer ${previous.attribs.timestamp} => ${file.timestamp}")
          true
        }
        case Some(previous) if previous.offset != file.ftpFile.getSize => {
          logger.info(s"${file.name} was not completely read & committed ${previous.offset} => ${file.ftpFile.getSize}")
          true
        }
        case _ =>
          logger.info(s" no fetch required")
          false
      }
    }
```
would become this : 

```

      metadata match {
        case None =>
          logger.info(s"${file.name} hasn't been seen before")
          true
        case Some(previous) if previous.attribs.size != file.ftpFile.getSize => {
          logger.info(s"${file.name} size changed ${previous.attribs.size} => ${file.size}")
          true
        }
        case Some(previous) if previous.attribs.timestamp != file.timestamp => {
          logger.info(s"${file.name} is newer ${previous.attribs.timestamp} => ${file.timestamp}")
          true
        }
        case Some(previous) if previous.offset != file.ftpFile.getSize => {
          logger.info(s"${file.name} was not completely read & committed ${previous.offset} => ${file.ftpFile.getSize}")
          true
        }
        case _ =>
          logger.info(s" no fetch required")
          false
      }
```

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -53,77 +55,155 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
         case None => FileMetaData(attributes, hash, Instant.now, Instant.now, Instant.now)
         case Some(old) => FileMetaData(attributes, hash, old.firstFetched, old.lastModified, Instant.now)
       }
       Option(prevFetch, FetchedFile(meta, bytes))
```
There seems to be some commonalities between the two branches of the if.

Mainly, what differs is how the `bytes` are retrieved. `meta` seems identical, `hash` and `attributes` as well.

How about having two functions for each of the `bytes` retrieval method ?

nicolas-coulon replied:
The intent here is to replace the legacy code by the new code... see the description of this pull request (just added)

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -53,77 +55,155 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+          logger.info(s"fetched ${current.meta.attribs.path}, bytes read= ${current.body.length} , from offset=0")
+          val nextOffsetToReadFrom = current.body.length
+          (current.meta.inspectedNow().offset(nextOffsetToReadFrom), FileBody(current.body, nextOffsetToReadFrom))
+      }
```
I'm having a very hard time following the ifs-in-ifs-in-matches. Could a suggest breaking up in function so the we can read separately the control flow logic and the actions taken in each branch ?

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpSourceConfig.scala:
```diff
@@ -65,6 +72,7 @@ object FtpSourceConfig {
     .define(FileMaxAge, Type.STRING, Importance.HIGH, "ignore files older than this; ISO8601 duration")
     .define(MonitorTail, Type.LIST, "", Importance.HIGH, "comma separated lists of path:destinationtopic; tail of file is tracked")
     .define(MonitorUpdate, Type.LIST, "", Importance.HIGH, "comma separated lists of path:destinationtopic; whole file is tracked")
+    .define(MonitorUpdateSlice, Type.LIST, "", Importance.HIGH, "comma separated lists of path:destinationtopic; whole file is tracked and teeated as slices")
```
typo: `teeated` => `treated`

I'm wondering if `slice` is the best way to name this new feature (both for its name and for its description). In my case, the name doesn't help me understand what it does and the description doesn't clarify it. Could we brainstorm this to try to find a meaningful name and description ?

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpSourceConfig.scala:
```diff
@@ -80,8 +88,10 @@ class FtpSourceConfig(props: util.Map[String, String])
-      getList(FtpSourceConfig.MonitorUpdate).asScala.map { case topicPathRegex(path, topic) => MonitorConfig(topic, path, tail = false) }
+    getList(FtpSourceConfig.MonitorTail).asScala.map { case topicPathRegex(path, topic) => MonitorConfig(topic, path, mode = MonitorMode.Tail) } ++
+      getList(FtpSourceConfig.MonitorUpdate).asScala.map { case topicPathRegex(path, topic) => MonitorConfig(topic, path, mode = MonitorMode.Update) } ++
+      getList(FtpSourceConfig.MonitorUpdateSlice).asScala.map { case topicPathRegex(path, topic) => MonitorConfig(topic, path, mode = MonitorMode.UpdateSlice) }
```
Proposition:

```scala
val modes = (FtpSourceConfig.MonitorTail,  FtpSourceConfig.MonitorUpdate, FtpSourceConfig.MonitorUpdateSlice)
modes.flatMap(mode => getList(mode).asScala.map { case topicPathRegex(path, topic) => MonitorConfig(topic, path, mode) })
```

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -53,77 +55,155 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+
       val meta = prevFetch match {
         case None => FileMetaData(attributes, hash, Instant.now, Instant.now, Instant.now)
         case Some(old) => FileMetaData(attributes, hash, old.firstFetched, old.lastModified, Instant.now)
```
Just to be sure: FileMetaData is what's stored in kafka right ? Which means this is where we store the offset in the file ?

jbruggem reviewed (commented):
Looking great so far ! I like it.

nicolas-coulon committed:
Remove "FtpSourceConfig.MonitorUpdateSlice" config in favor of optional config "FtpSourceConfig.MonitorSliceSize"


#328

nicolas-coulon committed:
add test to check "update mode" and "tail mode" when processing by slices + related bug fixes

#328

nicolas-coulon committed:
rearrange code in order to make the review easier

#462

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -87,13 +118,58 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+    //Trim end
+    val bytes = util.Arrays.copyOfRange(buffer, 0, bytesRead)
+    // it doesn't make sense to compute the hash for chunks/slices
+    val hash = DigestUtils.sha256Hex(bytes)
```
The comments says it doesn't make sense to compute the hash, but the code makes me believe it is actually computed nonetheless.
I have three questions, and I believe the answer to those questions should be added to this comment:
1. Why doesn't it make sense ?
2. Why is is calculated anyway (if indeed it is)
3. What would be the better thing to do wand why are we not doing it ?

nicolas-coulon replied:
Computing the hash code of a file was used by landoop to check if the file was changed. If we had to do the same, we would need to read the entire file. So I'd just remove that line.
To determine if a file has changed i look to the last modified date and size

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -87,13 +118,58 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+    val attributes = new FileAttributes(file.path, file.ftpFile.getSize, file.ftpFile.getTimestamp.toInstant)
+
+    val meta = prevFetch match {
+      case None => FileMetaData(attributes, hash, Instant.now, Instant.now, Instant.now)
```
Those three `Instant.now` will be slightly different. Would it be more coherent to only call that method once ?

nicolas-coulon replied:
Functionnaly it's not very important, landoop does the same. That data isn't used in any logic.

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -124,24 +200,56 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
 
+  def bySlicesHandleFetchedFile(mode: MonitorMode, prevFetch: Option[FileMetaData], current: FetchedFile): (FileMetaData, FileBody) = {
+    prevFetch match {
+      case Some(previous) if mode == MonitorMode.Update && (previous.attribs.timestamp != current.meta.attribs.timestamp || previous.attribs.size != current.meta.attribs.size) =>
```
All these case classes on `FileMetaData` objects are getting messing. 

My suggestion: add some methods to `FileMetaData` with what you want to check. 

For example if you add a `hasChangedSince` method, you could do:
```scala
case Some(previous) if mode == MonitorMode.Update && current.hasChangedSince(previous)
```

nicolas-coulon replied:
good idea

toch replied:
I think we have a good candidate for a strategy (this comment applies to other method whose the behavior depends on the value of the `mode`). Indeed, the applied behavior depends on the `mode`. I would create a strategy for each mode, and call it when the conditions are met.

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpSourceConfig.scala:
```diff
@@ -80,8 +88,8 @@ class FtpSourceConfig(props: util.Map[String, String])
-    getList(FtpSourceConfig.MonitorTail).asScala.map { case topicPathRegex(path, topic) => MonitorConfig(topic, path, tail = true) } ++
-      getList(FtpSourceConfig.MonitorUpdate).asScala.map { case topicPathRegex(path, topic) => MonitorConfig(topic, path, tail = false) }
+    getList(FtpSourceConfig.MonitorTail).asScala.map { case topicPathRegex(path, topic) => MonitorConfig(topic, path, mode = MonitorMode.Tail) } ++
+      getList(FtpSourceConfig.MonitorUpdate).asScala.map { case topicPathRegex(path, topic) => MonitorConfig(topic, path, mode = MonitorMode.Update) }
```
My proposition of using a flatMap here still stands.

nicolas-coulon replied:
I don't see how/ what benefit?

jbruggem replied:
Cleaner / more readable / more scala-ish.

Not critical, that's for sure.

jbruggem commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpSourceTask.scala:
```diff
@@ -59,17 +59,19 @@ class FtpSourcePoller(cfg: FtpSourceConfig, offsetStorage: OffsetStorageReader)
+    //val (head, tail) = stream.splitAt(cfg.maxPollRecords)
+    //buffer = tail
+    //head
+    fetchRecords()
```
why is the commented code not needed any more ?

nicolas-coulon replied:
Since we work by slices, why would we need a buffer?

nicolas-coulon replied:
but you are right, I can simply let the code as it is.

toch replied:
what normally do this polling method? polling until it gets enough new data in the buffer? what is a record here?

toch commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -43,7 +45,9 @@ object EmptyFileBody extends FileBody(Array[Byte](), 0)
 // instructs the FtpMonitor how to do its things
 case class FtpMonitorSettings(host:String, port:Option[Int], user:String, pass:String, maxAge: Option[Duration],
-                              directories: Seq[MonitoredPath], timeoutMs:Int, protocol: FtpProtocol, filter:String)
+                              directories: Seq[MonitoredPath], timeoutMs:Int, protocol: FtpProtocol, filter:String, sliceSize:Int){
```
to stay bckw compatible, I'd either give a default value to sliceSize of 0 or add a ctor with that option.

nicolas-coulon replied:
Bydefault it is -1 (when no config is provided), it is set in the FtpConnectorConfig.scala. So it is backward compatible

toch commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -87,13 +118,58 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+  def bySlicesFetch(mode: MonitorMode, file: AbsoluteFtpFile, prevFetch: Option[FileMetaData]): Option[(Option[FileMetaData], FetchedFile)] = {
+    logger.info(s"fetching ${file.path}")
+
+    val offsetToReadFrom = prevFetch match {
```
I'd extract that into a dedicated method.

toch commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -87,13 +118,58 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+
   // translates a MonitoredDirectory and previously known FileMetaData into a FileMetaData and FileBody
-  def handleFetchedFile(tail: Boolean, prevFetch: Option[FileMetaData], current: FetchedFile): (FileMetaData, FileBody) =
+  def handleFetchedFile(mode: MonitorMode, prevFetch: Option[FileMetaData], current: FetchedFile): (FileMetaData, FileBody) = {
```
for backward compatibility, it'd be good to keep the old signature and call the new one depending on the boolean value.

toch commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -124,24 +200,56 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+      case Some(previous) if mode == MonitorMode.Update && (previous.attribs.timestamp != current.meta.attribs.timestamp || previous.attribs.size != current.meta.attribs.size) =>
+        logger.info(s"distant file was modified since last fetch, previously timestamp=${previous.attribs.timestamp}, current timestamp=${current.meta.attribs.timestamp}")
+        val nextOffsetToReadFrom = current.body.length
+        (current.meta.inspectedNow().offset(nextOffsetToReadFrom), FileBody(current.body, 0))
```
why the parenthesis?

toch commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala:
```diff
@@ -124,24 +200,56 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+    prevFetch match {
+      case Some(previous) if mode == MonitorMode.Update && (previous.attribs.timestamp != current.meta.attribs.timestamp || previous.attribs.size != current.meta.attribs.size) =>
+        logger.info(s"distant file was modified since last fetch, previously timestamp=${previous.attribs.timestamp}, current timestamp=${current.meta.attribs.timestamp}")
+        val nextOffsetToReadFrom = current.body.length
```
is the `current.body.length`, the length of the whole file content? or does it correspond to the the slice length that have been read?

toch commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpSourceConfig.scala:
```diff
@@ -46,13 +52,14 @@ object FtpSourceConfig {
   val FileConverter = "connect.ftp.fileconverter"
   val SourceRecordConverter = "connect.ftp.sourcerecordconverter"
-  val FtpMaxPollRecords = "connect.ftp.max.poll.records"
+  //val FtpMaxPollRecords = "connect.ftp.max.poll.records"
```
to uncomment?

toch commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/MaxLinesFileConverter.scala:
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2017 Datamountaineer.
```
That's not right, it should be this year and ENX

toch commented on kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/MaxLinesFileConverter.scala:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  override def getFileOffset(path: String): Option[FileMetaData] = {
+    metaStore.get(path)
```
is it correct that we continue to use the in memory store to store and retrieve the offset?

nicolas-coulon replied:
I didn't touched to that. It isn't stored in a topic.

toch commented:
@nicolas-coulon what do you need to run the test?

nicolas-coulon committed:
add some adaptations after review.

* make hascode optional
* encapsulate 'file was changed' logic
* re-integrate 'max poll records'

nicolas-coulon commented:
> @nicolas-coulon what do you need to run the test?

In the directory `kafka-connect-ftp`, run `gradle test`

nicolas-coulon committed:
add some adaptations after review.

* extract logic to determine offset to read from in a dedicated function
* review copy right

nicolas-coulon committed:
Deactivate buffering if processing by slices

#462

nicolas-coulon committed:
Reconnect to ftp just before each read of slice

Remove ftp.completePendingCommand(): we only read data, no transactional support needed.

#462

nicolas-coulon committed:
Make ftp timeout configurable


#462

nicolas-coulon committed:
add description to some configuration parameters

#462

nicolas-coulon committed:
There seems to be a limit to the amount of bytes retrieved from FTP in one call => get the slice bytes by multiple reads.

#462

Antwnis commented:
How are you progressing with this PR ? Is it ready to be raised against stream-reactor ?!

nicolas-coulon commented:
> How are you progressing with this PR ? Is it ready to be raised against stream-reactor ?!

Sorry to reply so late to your question. We completely missed your reaction (holidays etc).

The fact is that we are using this PR in production for month so it is ready to be raised against stream-reactor

What steps should we follow to get this merged ? I guess you intend to make a code review first?

jbruggem committed:
remove unecessary copyright
